### ISSUE TITLE
chart doesn't fit in the device size (Ionic app)

### ISSUE THREAD

**Issue by saurabh47** [highcharts/highcharts-angular#149]
Hello guys,
I'm using Highcharts in my Ionic 4 app. I'm using the highcharts-angular library.
when an app initially gets loaded. the chart doesn't fit the device size. I have tried compiled app on my android device. the same issue is happening.
![ionic-highcharts](https://user-images.githubusercontent.com/15173289/60507847-b8e5f480-9ce6-11e9-832f-a18cab45db8b.PNG)
After resizing the screen chart fit screen properly.
please help me.
I have also tried to change the width & height property. But it doesn't work.

someone also mentioned the same issue:-
[https://stackoverflow.com/questions/55886989/example-highcharts-in-ionic-4?answertab=votes#tab-top](url)

My code:-
`<highcharts-chart
        style="display: flex;"
        [Highcharts]="Highcharts"
        [constructorType]="chartConstructor"
        [options]="chartOptions"
      ></highcharts-chart>`

Ionic:

   Ionic CLI                     : 5.0.2 (C:\Users\SaurabhGangamwar\AppData\Roaming\npm\node_modules\ionic)
   Ionic Framework               : @ionic/angular 4.6.0
   @angular-devkit/build-angular : 0.13.9
   @angular-devkit/schematics    : 7.3.9
   @angular/cli                  : 7.3.9
   @ionic/angular-toolkit        : 1.5.1

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : android 8.0.0
   Cordova Plugins   : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 4.1.1, (and 4 other plugins)

Utility:

   cordova-res : not installed
   native-run  : 0.2.7

    "highcharts": "^7.1.2",
    "highcharts-angular": "^2.4.0",

Here is my complete code
[https://github.com/saurabh47/charts-ionic](url)

**Comment by KacperMadej** [comment]
Hi @saurabh47 

Thank you for contacting us about the problem.

Your repository doesn't look like it contains Highcharts or Highcharts Angular wrapper.
https://github.com/saurabh47/charts-ionic/search?q=highcharts&unscoped_q=highcharts

The problem you have described looks like a problem with CSS and layout. You could check if style for the chart container is correctly set and in case a resize is done to the layout without triggering a resize event that would cause Highcharts chart to adjust to new container size you could call `chart.reflow()`.

If you will be able to provide a working demo, showing the problem, that I could debug then I could check the code and suggest a tested solution.

**Comment by saurabh47** [comment]
thanks for your reply.
sorry, forget to push changes to the repository .
please take look at my code.

I have added any custom CSS.

**Comment by saurabh47** [comment]
@KacperMadej 
Please help, I have been stuck on this problem from the past 2 days.
thanks

**Comment by KacperMadej** [comment]
The "display: flex;" is not allowing Highcharts code to get container width and height at the time of initial chart size calculations. This could be resolved by postponing the component or by setting the chart options  with a slight delay - long enough to wait for the layout and short enough to not be noticeable for the chart user. Just calling `chart.reflow()` in chart's callback with a minimal delay will be enough if you don't care about initial animation. If you care please see the below code:

```js
import { Component, OnInit } from '@angular/core';
import * as Highcharts from 'highcharts/highcharts.src.js';

@Component({
  selector: 'app-my-chart',
  templateUrl: './my-chart.component.html',
  styleUrls: ['./my-chart.component.scss'],
})
export class MyChartComponent implements OnInit {

  months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

  chart;
  updateFromInput = false;
  Highcharts = Highcharts;
  chartConstructor = 'chart';
  chartCallback = (chart) => {
    setTimeout(()=>{
      chart.reflow();
      this.chartOptions = {
        chart: {
          type: 'column',
        },
        title: {
          text: 'Fruit Consumption'
        },
        xAxis: {
          categories: this.months
        },
        yAxis: {
          title: {
            text: 'Fruit eaten'
          }
        },
        series: [{
          name: 'Jane',
          data: [1, 0, 4,1,2,3,6,7,8,4,1,6]
        }, {
          name: 'John',
          data: [1, 0, 4,1,2,3,6,7,8,4,1,6]
        }]
      };
    }, 0);
  };
  chartOptions = {};
  constructor() { }

  ngOnInit() {}

}
```

```html
<highcharts-chart
        style="display: flex;"
        [Highcharts]="Highcharts"
        [constructorType]="chartConstructor"
        [options]="chartOptions"
        [callbackFunction]="chartCallback"
        [oneToOne]=true
      ></highcharts-chart>
```

This case is not a bug and should be resolved through Highcharts support - mostly to avoid the delay that you have experienced.

I hope this resolves your problem - I have tested this in your app, so it should be fine.
Please let me know if you have any more questions related to the case or if we could close the issue. @saurabh47

**Comment by saurabh47** [comment]
thank you so much @KacperMadej 
It's as expected working.

**Comment by tuannd2207** [comment]
i don't understand why highcharts makes the chart fit the container by resizing the window. That is so fucking stupid. why don't you guys make it fit the container by the width of the container? why ?

**Comment by Alay2812USC** [comment]
Hey , I am facing the same issue @KacperMadej . But your solution isnt working for me

**Comment by kaiwangx** [comment]
Try lazy loading https://material.angular.io/components/tabs/overview#lazy-loading

**Comment by mconner** [comment]
I'm running into the same issue, though it only seems to happen under certain circumstances, (I'm guessing, due to timing).
I tried using the callbackFunction, as suggested, but I run into some other issues with my code expecting that the chart was initialized.  I could probably work through this, but the underlying issue seems to be that element at `highcharts-chart  .highcharts-container ` is setting the style to 600px by 400px (i.e.: some default).  Explicitly setting a style on the component does not seem to work, nor does calling reflow(), even with a 5 second delay. However, if I manually resize the containing element (this is in a dashboard widget, with resize handles), it sizes correctly. 

I found that by using ng-deep, I can work around this:
```
:host {
...
  &::ng-deep highcharts-chart .highcharts-container {
        flex: 1 1 auto;
   }
...
}
```
Or similarly using:  `width: 100% !important`, instead of flex  Either way, I'm basically overriding width and height styles explicitly being set on the element through scss.   Clearly, this is a expedient hack, and may have side effects.  In my case, I've a single chart in the enclosing element, so that simplifies things.

I  do understand tuannd2207's frustration (but, yeah, rude), but this seems to be more of a highcharts issue than a highcharts-angular issue, if I understand correctly.
